doc/man-pages/xbbc.1
.TH XBBC 1 04-Jan-2022 "Version 1.28" "X11-Basic"
.SH NAME
xbbc \- X11-Basic bytecode compiler 
.SH SYNOPSIS
.B xbbc
.RI "[ " option " ] " 
[\fB\-o\fR \fIoutfile\fR]
.IR <filename>

.SH DESCRIPTION

X11-Basic is a dialect of the BASIC programming language with graphics.

.B xbbc
can be used to make bytecode-binaries out of *.bas script files. These bytecode 
files can be interpreted by the X11-Basic virtual machine (\fIxbvm(1)\fR) 
and they run about 10-20 times faster than the interpreted .bas files. Since 
.B xbbc
is a real compiler, some restrictions to the code apply. (Not every X11-Basic 
program which does well on the interpreter can be correctly compiled),
.PP
For a desciption of the X11-Basic language and usage of the interpreter see
.I xbasic(1)\fR.

.SH Filenames

.B xbbc 
processes one input file. The suffix of the input file is usually .bas.
The default output file name is 
.I b.b
but you can specify alternate names with the -o option.

.SH OPTIONS
.TP
.BR \-o " \fIfile\fR
Place output in file 
.I file
\&. 
If you do not specify `\|\c
.B \-o\c
\&\|', the default is to produce a file `\|\c
.B b.b\c
\.
.TP
.BR \-\-version
shows version, program and other information and exits
.TP
.BR \-h ", " \-\-help
print a short help.

For detailed help and a description of the implemented basic commands take a look
at the README file which ships with the X11-Basic package.


.SH EXAMPLES
.nf
xbbc testme.bas
 produces a bytecode file b.b
xbbc testme.bas -o testme.b 
 produces a bytecode file testme.b
.fi

.SH The bytecode compiler

.B xbbc 
is a compiler. It compiles the BASIC source code into a bytecode of a
virtual stack machine. This virtual stack machine needs to be emulated on each
different operating system platform, but the bytecode itself is still portable.
The bytecode is very compact and similar to machine code, but still it is
independant of the actual processor hardware. With an additional small step it
can be compiled to native assembler or machine code. (For linux this is done by
the 
.I xb2c(1) 
translator and 
.I gcc(1)
, for other operating systems such a translator has not yet been created).

There is no way to recreate the source code from the bytecode. 

.SH VERSION
Page was created for V.1.18

.SH BUG REPORTS  

If you find a bug in the X11-Basic bytecode compiler, you should report it. But
first, you should make sure that it really is a bug, and that it appears in the
latest version of the X11-Basic package that you have.

Once you have determined that a bug actually exists, mail a bug report to
kollo@users.sourceforge.net. If you have a fix, you are welcome to mail that as
well! Suggestions may be mailed to the x11-basic bug tracking system.

Comments and bug reports concerning this manual page should be directed to
kollo@users.sourceforge.net.

.SH AUTHOR
Markus Hoffmann <kollo@users.sourceforge.net>

.SH COPYRIGHT
Copyright (C) 1997-2022 Markus Hoffmann <kollo@users.sourceforge.net>

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

.SH SEE ALSO
xbasic(1), xbvm(1), xbc(1), xb2c(1)


doc/HISTORY

Version history of X11-Basic
============================

In the beginning there were absurd plans, long sleepless nights, and incredible
boredom.

Very First Version (1990)
--------------------------
It was a Mailbox-Program with Basic features (only character-Graphics with
Vt100EMU) on my ATARI ST (!).

Version 0.1 First version with a version number    (1997)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- Mailbox-Program was rewritten in c on my Linux-PC
- first X11-Commands

Version 0.2 Alpha (1998)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- improved X11-Basic for use at the physics institute in Bonn
- Fixed last Problems of the Parser (didn't like -1*(4-3) )
- Added more X-features (redraw)
- First portation to a HP-Workstation
- Readline library in use

Version 1.00 Beta (Mar 1999)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- Makefile and configure script adapted from STonX
- True-color Displays
- X rootwindow, multiple windows
- INKEY$
- First GPL-Distribution
- FILESELECT,ALERT

Version 1.01 (Jun 1999)
~~~~~~~~~~~~~~~~~~~~~~
- POINT()

Version 1.02 (Jun 2000)
~~~~~~~~~~~~~~~~~~~~~~
- rsrc_load
- shell, daemon

Version 1.03 (Mar 2001)
~~~~~~~~~~~~~~~~~~~~~~~
- copyarea, sget, sput, get,put
- fill-Patterns
- drop down menues
- String-Variables may now contain nul-characters.
- POKE,DPOKE,LPOKE,EVEN(),ODD()
- LINK,UNLINK
- CVI(),CVD(),CVF(),CVL()
- MKI$(),MKD$(),MKL$(),MKF$()
- ON x GOSUB proc1,proc2,...
- ON x GOTO label1,label2,...
- INSTR(),RINSTR()

Version 1.04 (Apr. 2001)
~~~~~~~~~~~~~~~~~~~~~~~~
- splitted off shared library libx11basic.so (300kB)
- bug fixed which caused print a,b to segmentation fault
- DUMP "#"
- SYM_ADR()
- CALL, EXEC, EXEC()  with parameter list
- DUMP "K" DUMP "F"

Version 1.05 (Aug. 2001)
~~~~~~~~~~~~~~~~~~~~~~~~
- PRINT USING
- SYSTEM$()
- GLOB()
- RBOX, PRBOX
- EDIT, SAVE
- HELP
- new system errors
- started to translate everything into english

Version 1.06 (Jan. 2002)
~~~~~~~~~~~~~~~~~~~~~~~~

- included serial device support
- improved FILESELECT
- modified ALERT
- improved COLOR
- FORM_DIAL(), FORM_DO(), OBJC_DRAW()
- GPRINT

Version 1.07 (Aug. 2002)
~~~~~~~~~~~~~~~~~~~~~~~~
- Fehler mit Exponentialdarstellungen von Zahlen mit negativem Exponent
  ist nun endlich behoben !!!
- Anbindung an tine-Bibliothek und Kontrollsystem ist nun moeglich

Version 1.08 (Jan. 2003)
~~~~~~~~~~~~~~~~~~~~~~~~
- Support for tine-library
- a first MS-WINDOWS version is now available
- some new functions
- FIT and FIT_LINEAR
- SORT command
There was a bug introduced in the MENU command. Fixed it in release 1.08-4.

Version 1.09 (June 2003)
~~~~~~~~~~~~~~~~~~~~~~~~
- ANSI-Basic to X11-Basic converter included
- New example programs
- updated man-page and documentation
- new SOUND command
- EVAL, EVAL()
- modified serial line support
- modified the configure procedure
- some modifications for portability
- will now compile on FreeBSD and CYGWIN

Version 1.10 (Feb 2004)
~~~~~~~~~~~~~~~~~~~~~~~~
- new REPLACE$() function 
- improved speed 
- bug fixed in ROUND() function
- fixed bug in PUT_BITMAP
- fixed bug in serial device handling
- added sysvars COLS and ROWS
- new INODE/DEVICE/MODE/NLINK/SIZE file functions

Version 1.11 (Oct 2004)
~~~~~~~~~~~~~~~~~~~~~~~~
- Parser now accepts unique short commands
- fixed little bug in INPUT
- new chapters in manual
- new commands: TOPW, BOTTOMW, INFOW, FULLW, CLIP
- found bug in ON ERROR and ON BREAK
- new FILL command
- changed TAB treatement in sourcecode
- worked on TINE implementation
- worked on the ARRAY functions
- new SHL(), SHR() function

Version 1.12 (Mar 2005)
~~~~~~~~~~~~~~~~~~~~~~~~
- fixed bug with DEFMARK
- experimental LISTSELECT
- worked on MATRIX functions: SOLVE(),INV(),DET()
- OPEN "UU"  UDP-Sockets
- CONNECT, SEND, RECEIVE
- New String Functions: REVERSE$(), MTFE$(), MTFD$(), RLD$(), RLE$()
- CRC()
- New Code Engines: BETE$(), BWTD$(), COMPRESS$()

Version 1.13 (Feb 2006)
~~~~~~~~~~~~~~~~~~~~~~~~
- fixed bug in INPUT$()
- woked in the TINE implementation (TINELISTEN, TINEBROADCAST)
- new MKA$(), CVA()
- fixed bug in PAUSE/DELAY
- implemented ioctl() function

Version 1.14 (Apr 2007)
~~~~~~~~~~~~~~~~~~~~~~~~
- improved the init_program routines
- woked on the bytecode compiler
- new programs xbbc and xbvm  + man pages
- the bytecodecompiler and virtual machine are still full of bugs
- some sample programs can already be compiled to bytecode, 
  they run 4 to 10 times faster 
- Bug fixed which caused "Command  not allowed in interactive mode" errors
  This bug was introduced in Version 1.14 since release 1
- worked on the control system routines (DOOCS)
- fixed bug with fgetc
- crosscompiled it with arm-linux for embedded linux
- new CHDIR command
- new SPAWN command (does still not work as wanted)
- created doocsxbasic modification to use the FLASH control system at DESY


Release notes for Version 1.15 (Feb 2008)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- New command: PIPE
- New commands: GET_SCREENSIZE, GET_GEOMETRY
- Implemented BMP file format (for PUT)
- produced a framebuffer-graphics version of X11-basic


Release notes for Version 1.16 (Feb 2010)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- Bugfix in SORT (Feb 2010)
- new function word$() June 2010
- Improved DATA/RESTORE handling Dez 2010
- Improved xbbc/bytecode/xbvm
- Fixed memory leak accuring in varptr()
- added xbbc and xbvm to windows distribution
- started xb2c project
- new psydocompiler for MS-WINDOWS, can produce standalone .exe files 
- Bug fixed: LINEINPUT and LINEINPUT$() now can read arbitrary long lines
- fixed bug with PBOX: now 2nd coordinate is inside the filled area.
- new commands WATCH/(UNWATCH)/FILEEVENT$
- fixed bug with bitmaps (PUT)
- added support for bitmaps with transparency/alpha channel
- fixed bug in DEFFILL
- fixed bug in objc_draw()

Release notes for Version 1.17 (May 2011)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- Adapted windows version to use the SDL library (2011-04)

src/xbbc.c
/* XBBC.C   The X11-basic bytecode compiler         (c) Markus Hoffmann
*/

/* This file is part of X11BASIC, the basic interpreter for Unix/X
 * ======================================================================
 * X11BASIC is free software and comes with NO WARRANTY - read the file
 * COPYING for details
 */


#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <errno.h>
#include <string.h>
#include <strings.h>
#include <sys/types.h>
#include <unistd.h>
#if defined(__CYGWIN__) || defined(__MINGW32__)
#include <windows.h>
#endif

#if defined WINDOWS || defined ANDROID
#define EX_CANTCREAT	73	/* can't create (user) output file */
#define EX_NOINPUT	66	/* cannot open input */
#define EX_DATAERR      65      /* data format error */
#define EX_OK 0
#else
#include <sysexits.h>
#endif

#include "defs.h"
#include "x11basic.h"
#include "bytecode.h"

char ifilename[128]="new.bas";   /* default input file name */
static char *ofilename="b.b";    /* default output file name */
#ifdef ATARI
int compile_flags=COMPILE_DEFAULT|COMPILE_VERBOSE;
int verbose=1;
#else
int compile_flags=COMPILE_DEFAULT;
int verbose=0;
#endif
int loadfile=FALSE;
int dostrip=0;                   /* dont write symbol tables */
int programbufferlen=0;
char *programbuffer=NULL;
char **program=NULL;
int prglen=0;


static void intro(){
  puts("**************************************************\n"
       "*  X11-Basic bytecode compiler                   *\n"
       "*              by Markus Hoffmann 1997-" VERSION_YEAR " (c)  *");
  printf("* V." VERSION "/%04x      date:    " __DATE__ " " __TIME__ " *\n",BC_VERSION);
  puts("**************************************************\n");
}
static void usage(){
  printf("Usage:\n%s [-o <outputfile> -h] [<filename>]\t--- compile program\t[%s]\n\n","xbbc",ifilename);
  printf("-o <outputfile>\t--- put result in file\t[%s]\n",ofilename);
  puts("-h --help\t--- Usage\n"
       "-c\t\t--- also compile in comments\n"
       "-n\t\t--- also compile in nop's\n"
       "-s\t\t--- strip symbols\n"
       "-v\t\t--- be more verbose");
}


/* process command line  */

static void kommandozeile(int anzahl, char *argumente[]) {
  for(int count=1;count<anzahl;count++) {
    if(!strcmp(argumente[count],"-o")) ofilename=argumente[++count];
    else if(!strcmp(argumente[count],"-h")) {
      intro();usage();
    } else if(!strcmp(argumente[count],"-n")) compile_flags=compile_flags|COMPILE_NOOPS;
    else if(!strcmp(argumente[count],"-c"))   compile_flags=compile_flags|COMPILE_COMMENTS;
    else if(!strcmp(argumente[count],"-s"))   dostrip=!dostrip;
    else if(!strcmp(argumente[count],"-v"))   verbose++;
    else if(!strcmp(argumente[count],"-q"))   verbose--;
    else if(!strcmp(argumente[count],"--help")) {
      intro();usage();
    } else if(*(argumente[count])=='-') printf("%s: Unknown option %s ignored.\n",argumente[0],argumente[count]); 
    else {
      if(!loadfile) {
        loadfile=TRUE;
        strcpy(ifilename,argumente[count]);
      }
    }
  }
}

int main(int anzahl, char *argumente[]) {
  /* Initialize data segment buffer */
  if(anzahl<2) { /* Nothing to do, just display intro ... */
    intro();usage();
  } else {
    kommandozeile(anzahl, argumente);    /* process command line */
    if(loadfile) {
      if(exist(ifilename)) {
        libx11basic_init();  /* prepare everything, so that loadprg() can work. */
	#ifdef ATARI
	intro();
	printf("<-- %s\n",ifilename);
	#endif
        int prgerrors=loadprg(ifilename);
        if(prgerrors<0) {
          printf("%s: ERROR: program %s contains errors. Cannot compile.\n",argumente[0],ifilename);
	  exit(EX_DATAERR);
	} else if(prgerrors) {
          printf("%s: WARNING: program %s contains warnings. Will compile anyway.\n",argumente[0],ifilename);
	}
	
	/* Initialize compile block: */
	
	COMPILE_BLOCK cb;
	bzero(&cb,sizeof cb);
	if(verbose>0) compile_flags|=COMPILE_VERBOSE;
#ifdef ATARI
	printf("Compile ... \n");
#endif
	if(compile(&cb,compile_flags)<0) {
  	  printf("ERROR: compilation failed. Saving the result anyways.\n");
	}
	int ret=save_bytecode(ofilename,&cb,dostrip);
	if(ret==-1) exit(EX_CANTCREAT);
	#ifdef ATARI
	printf("done.\n");
	#endif
	free_cb(&cb);
      } else {
        printf("%s: ERROR: %s: file not found!\n",argumente[0],ifilename);
	#ifdef ATARI
	sleep(2);
	#endif
        exit(EX_NOINPUT);
      }
    } else exit(EX_NOINPUT);
  }
#ifdef ATARI
  sleep(4);
#endif
  return(EX_OK);
}
